resources/assets/lib/store.ts

```diff
@@ -0,0 +1,167 @@
+/**
+ *    Copyright (c) ppy Pty Ltd <contact@ppy.sh>.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Shopify from 'shopify-buy';
+import { toShopifyVariantId } from 'shopify-gid';
+
+declare global {
+  interface Window {
+    Store: Store;
+  }
+}
+
+// process.env.$ has to be static as it is injected by webpack at compile time.
+const options = {
+  domain: process.env.SHOPIFY_DOMAIN,
+  storefrontAccessToken: process.env.SHOPIFY_STOREFRONT_TOKEN,
+};
+
+const client = Shopify.buildClient(options);
+
+interface LineItem {
+  quantity: number;
+  variantId: string;
+}
+
+export class Store {
+  private static instance: Store;
+
+  static init() {
+    if (this.instance == null) {
+      this.instance = new Store();
+    }
+
+    return this.instance;
+  }
+
+  private constructor() {
+    $(document).on('click', '.js-store-checkout', this.beginCheckout.bind(this));
+    $(document).on('click', '.js-store-resume-checkout', this.resumeCheckout.bind(this));
+
+    $(document).on('turbolinks:load', () => {
+      $('.js-store-checkout').prop('disabled', false);
+    });
+
+    $('.js-store-checkout').prop('disabled', false);
+  }
+
+  async beginCheckout(event: Event) {
+    if (event.target == null) { return; }
+
+    const orderId = osu.presence((event.target as HTMLElement).dataset.orderId);
+    if (orderId == null) {
+      throw new Error('orderId is missing');
+    }
+
+    const { isValid, lineItems } = this.collectShopifyItems();
```

Comment:
looks like belongs to `beginShopifyCheckout` (up to the `beginShopifyCheckout`, with confusing `lineItems` name which are actually Shopify line items) and add something like `data-provider` to the button.

Or could even its own class.